ota_proxy/db.py

```diff
@@ -1,43 +1,57 @@
 import sqlite3
-from dataclasses import dataclass
+from dataclasses import make_dataclass
+from contextlib import contextmanager
 from pathlib import Path
 from threading import Lock
 
 import logging
+from typing import Any, Dict, List, Tuple
 
 from .config import config as cfg
 
 logger = logging.getLogger(__name__)
 logger.setLevel(cfg.LOG_LEVEL)
 
 
-@dataclass
-class CacheMeta:
-    url: str
-    hash: str
-    size: int
-    content_type: str
-    content_encoding: str
+class _CacheMetaMixin:
+    _cols = cfg.COLUMNS
 
-    def to_tuple(self) -> tuple:
-        return (
-            self.url,
-            self.hash,
-            self.size,
-            self.content_type,
-            self.content_encoding,
-        )
+    def to_tuple(self) -> Tuple[Any]:
+        return tuple([getattr(self, k) for k in self._cols])
+
+    @classmethod
+    def row_to_meta(cls, row: Dict[str, Any]):
+        if not row:
+            return
+
+        res = cls()
+        for k in cls._cols:
+            setattr(res, k, row[k])
+
+        return res
+
+
+def make_cachemeta_cls(name: str):
+    # set default value of each field as field type's zero value
+    return make_dataclass(
+        name,
+        [(k, v.col_type, v.col_type()) for k, v in cfg.COLUMNS.items()],
+        bases=(_CacheMetaMixin,),
+    )
+
+
+CacheMeta = make_cachemeta_cls("CacheMeta")
```

Comment:
I can understand that you don't want to treat COLUMNS members in CacheMeta and OTACacheDB.
But this change reduces code readability slightly, I think.
Easy to understand is one of the important things.
I'd like to hear your opinion.

Reply:
Hard-coding the table column definition may increase the maintenance complexity.
If I the definition of table change(even the order of the column changed), not only the database operation related code needed to be adjusted, the `CacheMeta` class, and even the code that use `CacheMeta` may need to be adjusted.

Considering the above situation, I decided to dynamically create `CacheMeta` class directly from the db table definiton. Actually it is a simple version of ORM, since our requirement is not so complicated, I directly write one based on dataclass instead of relying on third-party module.